src/ui/controllers/booking/booking.ts

```diff
@@ -216,88 +219,121 @@ export async function renderOuch(
-		errors.push('errors.po-too-long')
-	}
+	if (callOffPo) {
+		session.payment = {
+			type: 'PURCHASE_ORDER',
+			value: `Call off: ${callOffPo.id}`,
+		}
+		session.save(() => {
+			res.redirect(`${req.originalUrl}/confirm`)
+		})
+	} else {
+		const organisation = await registry.follow(config.REGISTRY_SERVICE_URL,
+			['organisations', 'search', 'findByDepartmentCode'],
+			{ departmentCode: user.department }) as any
 
-	if (trimmed.match(/[#;.%]/)) {
-		errors.push('errors.po-special-characters')
+		if (!organisation) {
+			res.redirect('/profile')
+		} else {
+			res.send(
+				template.render('booking/payment-options', req, res, {
+					errors: req.flash('errors'),
+					paymentMethods: organisation.department.paymentMethods,
+					values: req.flash('values')[0] || (session.payment ? { [session.payment.type]: session.payment.value } : {}),
+				})
+			)
+		}
 	}
 
-	return errors
 }
 
-export function enteredPaymentDetails(
+export async function enteredPaymentDetails(
 	req: express.Request,
 	res: express.Response
 ) {
 	const session = req.session!
-
-	if (req.body['purchase-order']) {
-		const poErrors = validatePurchaseOrder(req.body['purchase-order'])
-
-		if (poErrors.length) {
-			poErrors.map((error: string) => {
-				req.flash('purchaseOrderErrors', req.__(error))
-			})
-			session.save(() => {
-				res.redirect(`${req.originalUrl}`)
-			})
-			return
-		} else {
-			session.po = req.body['purchase-order'].trim()
-			session.save(() => {
-				res.redirect(`${req.originalUrl}/confirm`)
-			})
-			return
+	session.payment = null
+
+	const user = req.user as model.User
+	const organisation = await registry.follow(config.REGISTRY_SERVICE_URL,
+		['organisations', 'search', 'findByDepartmentCode'],
+		{ departmentCode: user.department }) as any
+
+	let errors: string[] = []
+
+	for (const paymentMethod of organisation.department.paymentMethods) {
+		if (req.body[paymentMethod]) {
+			errors = validate(paymentMethod, req.body[paymentMethod])
+			if (!errors.length) {
+				session.payment = {
+					type: paymentMethod,
+					value: req.body[paymentMethod].trim(),
+				}
+			}
+			break
 		}
 	}
 
-	//TODO: REF LPFG-315 Financial approver booking flow was not updated
-	if (
-		req.body['financial-approver'] &&
-		/^\S+@\S+$/.test(req.body['financial-approver'])
-	) {
-		session.fap = req.body['financial-approver']
+	if (!session.payment && !errors.length) {
+		errors.push('errors.empty-payment-method')
+	}
+
+	if (errors.length) {
+		errors.map((error: string) => {
+			req.flash('errors', req.__(error))
+		})
+		req.flash('values', req.body)
 		session.save(() => {
-			res.redirect(`${req.originalUrl}/confirm`)
+			res.redirect(`${req.originalUrl}`)
 		})
 	} else {
-		req.flash('fapErrors', 'Not valid email address')
 		session.save(() => {
-			res.redirect(`${req.originalUrl}`)
+			res.redirect(`${req.originalUrl}/confirm`)
 		})
-		return
 	}
 }
 
+export function validate(type: string, po: string): string[] {
```

Comment:
MAINTAINABILITY - we should be thinking of moving this logic to the necessary services (in the future)